### PR TITLE
#197/webapp improvements

### LINKED ISSUE #197: Fixes and improvements for MVP 0.1
This task is about reviewing some detected issues as well as making some improvements in our mvp:

- **Improve search**
  - use a single search context so we don't have to pass states and callbacks around and we centralize everything that matters in a search: query and results
  - fix visual "jumps" each time we update the seach query, try to make results appear smoother.
  - make rows and header fully scrollable so we have more space in a smartphone screen.

- **Move dispatch page into a template:**
  - I read about [atoms, molecules, templates and pages](https://blog.kamathrohan.com/atomic-design-methodology-for-building-design-systems-f912cf714f53) and loved the idea! 
  - The idea is very similiar to the dispatch page but, to convert it into a more reusable template so we can easily use it in chats, games etc. Its main responsibility is to draw data but does not care about how it get it, that should be a responsibility of th page instantiating it.
  - usersPage now holds the responsability of handling data and pass it to templates. The same idea can be easily replicated with other types of rows pages like chats, games, groups etc.

- **Improve UX of editAvatar page and component:**
  -  Make editableAvatar work in touch screens
  -  Add feedback messages to know an action was successful
  - Make selected avatar appear and disabling editing while uploading.

### PR DESCRIPTION
**Pull request by vdedios**
closes #197 

:warning:  **(Read the issue so you get the full context of what has been made)** :warning: 

Since many changes where related I decided to include everything in a single PR. Play around with it and if you need help don't hesitate to ask :)

### CODE REVIEW

File: webapp/src/shared/components/Row/Row.tsx [hunk 1]
```diff
@@ -113,3 +119,11 @@ export default function Row({
     )) || <div className={`row paragraph-regular`}>{RowChildren}</div>
   );
 }
+
```
Comment (Damrod) [inline webapp/src/shared/components/Row/Row.tsx:122]: ??

Comment (vdedios) [reply webapp/src/shared/components/Row/Row.tsx]: leftovers :eyes:

File: webapp/src/shared/components/Input/Input.tsx [hunk 2]
```diff
@@ -46,6 +46,7 @@ export default function Input({
           onChange={onChange}
           name={name}
           type={type}
+          autoComplete="off"
```
Comment (Damrod) [inline webapp/src/shared/components/Input/Input.tsx:49]: I saw chrome complaining about the password not having auto complete off, but generally, if not for passwords, I don't think it's a problem to have auto complete on. Why do you disagree?

File: webapp/src/shared/components/templates/RowsPageTemplate/RowsPageTemplate.css [hunk 3]
```diff
@@ -1,7 +1,8 @@
-.dispatch-page {
```
Comment (Damrod) [inline webapp/src/shared/components/templates/RowsPageTemplate/RowsPageTemplate.css:1]: If you change the name and the folder of the jsx, maybe you should also change the name and folder of the css?

File: webapp/src/shared/components/templates/RowsPageTemplate/RowsPageTemplate.tsx [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+import SearchForm from '../../Input/SearchForm';
+import RowsList, { RowItem } from '../../RowsList/RowsList';
+import NavigationBar from '../../NavigationBar/NavigationBar';
+import { useData } from '../../../hooks/UseData';
+import Loading from '../../Loading/Loading';
+import { usersApi } from '../../../services/ApiService';
+import { useSearchContext } from '../../../context/SearchContext';
+
+type RowsPageTemplateProps<T> = {
+  dataValidator: (data: T) => boolean;
```
Comment (Damrod) [inline webapp/src/shared/components/templates/RowsPageTemplate/RowsPageTemplate.tsx:15]: The data validator is there to look at whether the type of the object is T. I don't think it makes sense to have the data validator take T as an argument. It's a runtime check.

File: webapp/src/shared/components/templates/RowsPageTemplate/RowsPageTemplate.css (hunk 3 above)
Comment (vdedios) [reply webapp/src/shared/components/templates/RowsPageTemplate/RowsPageTemplate.css:1]: I think the name of the folder is already changed. Maybe you meant the name of the base class to `rows-page-template`?

File: webapp/src/shared/components/Input/Input.tsx (hunk 2 above)
Comment (vdedios) [reply webapp/src/shared/components/Input/Input.tsx:49]: Yes, passwords was the main reason... and also, for me, usually is a bit annoying to have the autocomplete display all the crap you've already typed before :sweat_smile:

File: webapp/src/shared/components/templates/RowsPageTemplate/RowsPageTemplate.tsx [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+  const { result, fetchMoreResults } = useSearchContext();
+
+  const data = ((array: object): RowItem[] | null => {
+    let rows = [];
+
+    if (!Array.isArray(array)) {
+      return null;
+    }
+    for (let i = 0; i < array.length; i++) {
+      if (!dataValidator(array[i])) {
```
[the remark this answers is not in the document]
Comment (vdedios) [reply webapp/src/shared/components/templates/RowsPageTemplate/RowsPageTemplate.tsx:32]: agree, is sadly less readable to use a for loop than a map but, since we are sanitizing arrays with no limits in length, I think we gain a bit in performance saving an iteration. A web application could run on the latest iphone or on a 8 years old phone, depending on the device that could be expensive

File: webapp/src/shared/components/templates/RowsPageTemplate/RowsPageTemplate.css (hunk 3 above)
Comment (Damrod) [reply webapp/src/shared/components/templates/RowsPageTemplate/RowsPageTemplate.css:1]: Ah, I didn't notice the name change. No problem then.

File: webapp/src/shared/components/templates/RowsPageTemplate/RowsPageTemplate.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (Damrod) [reply webapp/src/shared/components/templates/RowsPageTemplate/RowsPageTemplate.tsx:32]: I have to say that we added the type validator before we had properly generated Api files. We didn't have proper type checking from the request methods. I think now the generated fetching methods provide runtime type checking. Maybe it's obsolete, and we don't really need this here at all anymore.

File: webapp/src/shared/components/templates/RowsPageTemplate/RowsPageTemplate.tsx [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+import NavigationBar from '../../NavigationBar/NavigationBar';
+import { useData } from '../../../hooks/UseData';
+import Loading from '../../Loading/Loading';
+import { usersApi } from '../../../services/ApiService';
+import { useSearchContext } from '../../../context/SearchContext';
+
+type RowsPageTemplateProps<T> = {
+  dataValidator: (data: object) => boolean;
+  dataMapper: (data: T) => RowItem;
+};
```
Comment (Damrod) [inline webapp/src/shared/components/templates/RowsPageTemplate/RowsPageTemplate.tsx:17]: I added the `button ?: JSX.Element` to what we used to call DispatchPage Props because even back then we could already see that we may need a button below the list of rows, maybe not for the users page, but certainly for the chat. Why did you decide to remove it, now that it looks like we may actually need this feature?

File: webapp/src/shared/components/Row/Row.tsx [hunk 7]
```diff
@@ -95,9 +99,9 @@ export default function Row({
         {RowChildren}
       </label>
       <input
+        type="file"
```
Comment (Damrod) [inline webapp/src/shared/components/Row/Row.tsx:102]: why did you move this three lines above? 😯

File: webapp/src/shared/components/Avatar/EditableAvatar.tsx [hunk 8]
```diff
@@ -37,7 +30,11 @@ export default function EditableAvatar({
     )}px`,
   };
   return (
-    <div className="editable-avatar">
+    <div
+      className={`editable-avatar editable-avatar--${
```
Comment (Damrod) [inline webapp/src/shared/components/Avatar/EditableAvatar.tsx:34]: what do you disable here? 😯

File: webapp/src/pages/UsersPage/UsersPage.tsx [hunk 9]
```diff
@@ -9,9 +9,12 @@ import {
   User,
   UserControllerGetUsersRequest,
 } from '../../shared/generated';
-import { DispatchPage } from '../../shared/components/index';
+import { RowsTemplate } from '../../shared/components/index';
 import { useCallback } from 'react';
 import { usersApi } from '../../shared/services/ApiService';
+import { SearchContextProvider } from '../../shared/context/SearchContext';
+
+const ENTRIES_LIMIT = 15;
```
Comment (Damrod) [inline webapp/src/pages/UsersPage/UsersPage.tsx:17]: why 15?

File: webapp/src/pages/EditAvatarPage/EditAvatarPage.tsx [hunk 10]
```diff
@@ -58,18 +67,32 @@ export default function EditAvatarPage() {
           avatarY: FormatNumber(picturePosition.y),
```
Comment (Damrod) [inline webapp/src/pages/EditAvatarPage/EditAvatarPage.tsx:67]: Maybe we should do this transformation inside useDrag, passing FormatNumber as a callback as well, or maybe we shouldn't pass reverseTransform as a callback here, and do both transformations here. This is something I was thinking of doing. What do you think?

File: webapp/src/shared/components/Avatar/EditableAvatar.css [hunk 11]
```diff
@@ -20,11 +25,10 @@
 }
 
 .editable-avatar__image {
-  user-drag: none;
   -webkit-user-drag: none;
   margin: 0;
-  width: 16rem;
-  height: 16rem;
+  width: 15rem;
```
Comment (Damrod) [inline webapp/src/shared/components/Avatar/EditableAvatar.css:30]: This is veeeeery tricky to change. But I understand where is this coming from. I'm not sure what's wrong with my parameters. I've fixed them twice, and I think there is, again, some problem with the scale of the image and the offset of the picturePosition, for each of the different avatar sizes. Maybe someone has some idea.

**Comment by Damrod** [comment]
Besides the comments I have left, and besides the fact that I haven't yet had the opportunity to actually run to see that it's working, I love it. Will take a look tomorrow. Thank you!

File: webapp/src/shared/hooks/UseIsElementVisible.tsx [hunk 12]
```diff
@@ -0,0 +1,31 @@
+import * as React from 'react';
+
+export function UseIsElementVisible() {
+  const [isVisible, setIsVisible] = React.useState(false);
+  const observer = React.useRef<IntersectionObserver | null>();
```
Comment (gbudau) [inline webapp/src/shared/hooks/UseIsElementVisible.tsx:5]: praise: Now we can reuse it in more places, good job :+1: 

suggestion: I think the convention for custom hooks is to start with lowercase letters -> `useIsElementVisible`?

**Review by gbudau — APPROVED** [review]
Thanks for improving it! LGTM :rocket:

File: webapp/src/pages/UsersPage/UsersPage.tsx (hunk 9 above)
Comment (fizcamposancos) [reply webapp/src/pages/UsersPage/UsersPage.tsx:17]: If I'm not wrong we already have this limit in the back-end set to 20, in the [user.pagination.dto.ts](https://github.com/42AGV/ft_transcendence/blob/main/transcendence-app/src/user/dto/user.pagination.dto.ts) file
MAX_USER_ENTRIES_PER_PAGE = 20;
Though, I am moving this constant to /shared/constants.tsx file with the chat limit that I am working on

File: webapp/src/pages/EditAvatarPage/EditAvatarPage.tsx [hunk 13]
```diff
@@ -105,15 +134,26 @@ export default function EditAvatarPage() {
...
+        <Text
+          variant={TextVariant.PARAGRAPH}
+          color={
+            status.type === 'success' ? TextColor.ONLINE : TextColor.OFFLINE
+          }
+        >
+          {status.message}
+        </Text>
+      )}
       <Button
```
Comment (Damrod) [inline webapp/src/pages/EditAvatarPage/EditAvatarPage.tsx:157]: The button is too wide: this is not how it was supposed to look in figma. Someone broke the button width at some point in early September.

File: webapp/src/shared/components/templates/RowsPageTemplate/RowsPageTemplate.tsx (hunk 6 above)
Comment (fizcamposancos) [reply webapp/src/shared/components/templates/RowsPageTemplate/RowsPageTemplate.tsx:17]: I agree with @Damrod, or at least I was using this button feature in the chat page.
Or we can add it separately afterwards

File: webapp/src/shared/components/templates/RowsPageTemplate/RowsPageTemplate.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (vdedios) [reply webapp/src/shared/components/templates/RowsPageTemplate/RowsPageTemplate.tsx]: Agree, data validation is not needed anymore, removed and simplified a bit this method!

File: webapp/src/shared/components/templates/RowsPageTemplate/RowsPageTemplate.tsx (hunk 6 above)
Comment (vdedios) [reply webapp/src/shared/components/templates/RowsPageTemplate/RowsPageTemplate.tsx:16]: I'd say to add later it in chat's page task, we will know better the needs we have there!

File: webapp/src/shared/components/Row/Row.tsx (hunk 7 above)
Comment (vdedios) [reply webapp/src/shared/components/Row/Row.tsx:102]: I think this is a leftover of me trying to add support for adding a local directory path to update the image, but it was tricky and unnecessary at the end

File: webapp/src/pages/UsersPage/UsersPage.tsx (hunk 9 above)
Comment (vdedios) [reply webapp/src/pages/UsersPage/UsersPage.tsx:13]: It is down the max_limit. This constant is necessary so we can set de offset of our pagination queries (we fetch data ENTRIES_LIMIT by ENTRIES_LIMIT)

File: webapp/src/pages/EditAvatarPage/EditAvatarPage.tsx (hunk 10 above)
Comment (vdedios) [reply webapp/src/pages/EditAvatarPage/EditAvatarPage.tsx:67]: I think just passing the transformed values to `useDrag` would be the best :+1:

File: webapp/src/shared/components/Avatar/EditableAvatar.css (hunk 11 above)
Comment (vdedios) [reply webapp/src/shared/components/Avatar/EditableAvatar.css]: Yes, this was me playing around with sizes, I have restored the values. Thanks for pointing out

File: webapp/src/shared/components/Row/Row.css [hunk 14]
```diff
@@ -17,14 +19,23 @@
   display: none;
 }
 
+.row-icon {
+  grid-column: 1/2;
+}
+
```
[the remark this answers is not in the document]
Comment (vdedios) [reply webapp/src/shared/components/Row/Row.css:25]: I think this should be `row-avatar`, we don't have any place indication its grid-column

File: webapp/src/shared/components/Row/Row.css [hunk 15]
```diff
@@ -17,14 +19,23 @@
...
 .row_text_wrapper,
 .row_text_wrapper > * {
   margin: 0;
   text-align: left;
+  overflow: hidden;
+  grid-column: 2/3;
```
[the remark this answers is not in the document]
Comment (vdedios) [reply webapp/src/shared/components/Row/Row.css:31]: good catch! updated to `grid-column: span 1;`

File: webapp/src/pages/EditAvatarPage/EditAvatarPage.tsx [hunk 16]
```diff
@@ -82,6 +105,12 @@ export default function EditAvatarPage() {
     setImgData({
       imgFile: event.target.files[0],
       imgName: event.target.files[0].name,
+      imgSrc: URL.createObjectURL(event.target.files[0]),
+    });
+    console.log(imgData);
+    setStatus({
+      type: 'pending',
+      message: '',
```
[the remark this answers is not in the document]
Comment (vdedios) [reply webapp/src/pages/EditAvatarPage/EditAvatarPage.tsx:113]: yes! again, leftovers :laughing:

File: webapp/src/shared/context/SearchContext.tsx [hunk 17]
```diff
@@ -0,0 +1,135 @@
...
+}
+
+// https://reactjs.org/docs/context.html#reactcreatecontext
+const SearchContext = React.createContext<Context<any> | undefined>(undefined);
+
+export const SearchContextProvider = <T,>({
+  children,
+  fetchFn,
+  maxEntries,
+}: SearchContextProps<T>): JSX.Element => {
```
[the remark this answers is not in the document]
Comment (vdedios) [reply webapp/src/shared/context/SearchContext.tsx:78]: I won't go with `query.limit`, its really and offset representing where in the data we are and where we should start fetching in the next interval

File: webapp/src/pages/EditAvatarPage/EditAvatarPage.tsx (hunk 13 above)
Comment (vdedios) [reply webapp/src/pages/EditAvatarPage/EditAvatarPage.tsx:156]: Yes not 1:1 like Figma :smile: Since this is no blocking to merge this, we can review it in future tasks/chat iterations

**Comment by vdedios** [comment]
Merging this! Thanks all of you for your comments :heart:

File: webapp/src/shared/hooks/UseIsElementVisible.tsx (hunk 12 above)
Comment (vdedios) [reply webapp/src/shared/hooks/UseIsElementVisible.tsx]: done!
